service/shadowsocks.go

```diff
@@ -0,0 +1,49 @@
+// Copyright 2024 The Outline Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package service
+
+import (
+	"errors"
+	"net/netip"
+	"time"
+
+	"github.com/Jigsaw-Code/outline-sdk/transport/shadowsocks"
+)
+
+type DebugLoggerFunc func(tag string, template string, val interface{})
+
+// findShadowsocksAccessKey implements a trial decryption search. This assumes that all ciphers are AEAD.
+func findShadowsocksAccessKey(clientIP netip.Addr, bufferSize int, src []byte, cipherList CipherList, logDebug DebugLoggerFunc) (*CipherEntry, []byte, time.Duration, error) {
+	// We snapshot the list because it may be modified while we use it.
+	ciphers := cipherList.SnapshotForClientIP(clientIP)
+
+	unpackStart := time.Now()
+	// To hold the decrypted chunk length.
+	chunkLenBuf := make([]byte, bufferSize)
+	for ci, elt := range ciphers {
+		entry := elt.Value.(*CipherEntry)
+		buf, err := shadowsocks.Unpack(chunkLenBuf, src, entry.CryptoKey)
```

Comment:
I believe this is broken. src should have the exact size `firstBytes[:cryptoKey.SaltSize()+2+cryptoKey.TagSize()` for TCP and be the full packet for UDP.

We should have a test to catch that. Try different ciphers.

I remember trying to unify TCP and UDP, but they are different. I believe we are better off keeping the logic separate. It makes the signature of each method simpler, rather than a more complex API you have now. It's easier to use it wrong, vs the current code.